Bump minimum supported Python version to 3.8

jamesobutler opened the pull request:
Per https://github.com/numba/numba/blob/aaa6a0099ac217959126d195ab43d37000f9624a/CHANGE_LOG#L5-L6

and [NEP29's drop schedule](On Dec 26, 2021 drop support for Python 3.7 (initially released on Jun 27, 2018))
`On Dec 26, 2021 drop support for Python 3.7 (initially released on Jun 27, 2018)`

this PR executes on dropping support for Python versions prior to 3.8.

<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
forum https://numba.discourse.group/.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities here, please click the arrow besides
   "Create Pull Request" and choose "Create Draft Pull Request".
   When it's ready for review, you can click the button "ready to review" near
   the end of the pull request
   (besides "This pull request is still a work in progress".)
   The maintainers will then be automatically notified to review it.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on main/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against main they should
   be resolved by merging main into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

jamesobutler commented:
@esc Thoughts on what you would like to do to handle the failing (Linux py38_np118_32bit) test which is due to no 32-bit version available for Python 3.8 from anaconda?

jamesobutler commented:
Following the successful merging of this PR, tasks such as those mentioned in https://github.com/numba/numba/issues/7917 can be followups.

esc commented:
> @esc Thoughts on what you would like to do to handle the failing (Linux py38_np118_32bit) test which is due to no 32-bit version available for Python 3.8 from anaconda?

it is likely that we will drop linux-32 and win-32 from the list of officially supported platforms

jamesobutler commented on azure-pipelines.yml:
```diff
@@ -29,15 +29,15 @@ jobs:
     matrix:
-      py37_np118_32bit:
+      py38_np118_32bit:
         # 32 bit linux only has np 1.15
-        PYTHON: '3.7'
+        PYTHON: '3.8'
         NUMPY: '1.18'
         CONDA_ENV: azure_ci
         BITS32: yes
         TEST_START_INDEX: 2
```
@esc Would you like me to go ahead and proceed with removing this 32-bit job in the matrix?

re [failure](https://dev.azure.com/numba/numba/_build/results?buildId=12756&view=logs&jobId=5f38b89c-308c-5e22-4982-c0a0f2a8cf5c&j=5f38b89c-308c-5e22-4982-c0a0f2a8cf5c&t=29dd518e-3410-5392-c104-01ec46cfe18b)

esc replied:
I am not sure yet, I am hoping to discuss this during the developer meeting today. Hopefully, we can find a consensus soon.

gmarkall replied:
Why was the job called `py37_np118_32bit` if the comment says `32 bit linux only has np 1.15`? (How could it have been working / passing tests, given that there will be differences in implementation between Numba and NumPy - was it just getting lucky and not hitting tests in which there would have been a difference?)

stuartarchibald replied:
I think that comment is from a time when `linux-32` conda packages were available, NumPy 1.15 being the last version. NumPy 1.18 comes as a wheel:
https://github.com/numba/numba/blob/fc28b2974904c0e1e02b62a9cf37f75023cfc80d/buildscripts/incremental/setup_conda_environment.sh#L84-L87

gmarkall replied:
Thanks for the clarification. I think the comment might as well be removed - it could be updated to say that it only has/had 1.15 conda packages available, but I'm not sure it's useful information to have here.

jamesobutler replied:
Ok yes I was a bit confused by that code comment as well.

I see why the 32bit CI job is failing now. There are no 32bit conda packages for supporting pip/gitpython/pyaml for Python 3.8, not just NumPy. They would all have to come as whl files from PyPI by pip installing them.

Should this conda env job still be run if there aren’t going to be 32bit Linux conda packages to use?

esc replied:
> I see why the 32bit CI job is failing now. There are no 32bit conda packages for supporting pip/gitpython/pyaml for Python 3.8, not just NumPy. They would all have to come as whl files from PyPI by pip installing them.

This is correct. In fact, IIRC, the last Python version to be supported for linux-32 via the Anaconda Distro Team is Python 3.7 -- so we don't even have a way to obtain a Python for this platform.

> Should this conda env job still be run if there aren’t going to be 32bit Linux conda packages to use?

Probably not, no. As mentioned previously: I hope to get consensus on this during todays developer meeting. As soon as we agree, we can give you instructions on how to proceed.

esc replied:
@jamesobutler we decided to remove support for the following come 0.57 (current estimated release date: Jan 2023):

* Python below 3.8
* Numpy below 1.19
* All 32 bit architectures

However, since the stencil tests depend on the 3.7 AST, we need to get the refactor #8325 in, before we can remove 3.7 testing from the public CI matrix. You can go ahead and remove the 32 bit test for now, but the merge of this PR will delayed a little.

jamesobutler replied:
@esc I saw that you approved #8325 and that it is ready to merge. I have gone ahead and pushed another commit to this PR that removes the 32bit testing in the CI. All checks have now passed 🎉. Let me know if you would like additional changes here in this PR. Here to help!

esc replied:
@jamesobutler yes, thank you very much. #8325 has now been merged and I hope to look at this one and #8318 and #7920 early next week. 🙏

esc commented:
@jamesobutler thank you for submitting this. After an OOB conversation with @stuartarchibald we came to the conclusion, that we want to stall this PR until the 0.56.1 release is out. We want to be cautious of cherry-picks and messing with the public CI config (azure). Since there is no immediate pressure to get this onto `main` that should be OK. I've got this one added to my tracker and will pick this up again post 0.56.1 release.

jamesobutler commented:
Will also need to address the following in this PR which drops Python 3.7:
https://github.com/numba/numba/blob/64c06beb0551117efc3b3687d9d66734e391f603/numba/core/utils.py#L379-L387

esc commented:
@jamesobutler now that 0.56.2 is out, I'd like to get back to this PR. It seems to now be in conflict with `main`. Since I haven't started reviewing it yet, would you be so kind as to rebase this onto the current `HEAD`. That would be great, thank you.

jamesobutler commented:
@esc I've rebased to fix merge conflicts. https://github.com/numba/numba/pull/7920 should probably be finalized and merged first as it removes the currently outdated code. I'm not the author of that PR so I won't be able to help move that along.

gmarkall commented:
@jamesobutler Could you please avoid rebasing as it messes up the review flow on Github (it gets confused and can't track changes between reviews) - could you please use merges instead?

gmarkall commented:
> @jamesobutler Could you please avoid rebasing as it messes up the review flow on Github (it gets confused and can't track changes between reviews) - could you please use merges instead?

Oops, sorry, @esc just pointed to me that he suggested the rebase because it wasn't reviewed yet. Please ignore the above!

jamesobutler commented:
Now that there are more merge conflicts since I originally fixed them up, how am I allowed to proceed here?

I know there is a merge conflict due to a change in my first commit `Bump minimum supported Python version to 3.8` and there is also a merge conflict due to changes in my second commit `Remove 32bit testing in CI`. I would normally rebase against the latest `main` branch and fix up each of my two commits and force push. That way the changes are maintained cleanly when reviewing what is needed specifically for removing Python 3.7 support and then when specifically looking at the changes needed for removing 32-bit testing in CI. Then when this PR is merged both commits can be integrated into the `main` branch with clean history of what was changed in each.

gmarkall commented:
> Now that there are more merge conflicts since I originally fixed them up, how am I allowed to proceed here?

I think it's still OK to rebase here - I think its just my comment from earlier that caused confusion, apologies.

jamesobutler commented:
@esc and @gmarkall I’ve rebased again to fix up merge conflicts and the CI is green.

jamesobutler commented:
Rebased now that https://github.com/numba/numba/pull/7920 is merged. So everything below python 3.7 should be gone by now.

esc commented:
@jamesobutler thank you for the updates, I have submitted this to the Anaconda Internal Build Farm to smoketest (i.e. a short test, only a few hours) it. Since this is such a sweeping change I'd like to give it a full run (longer test, runs 12 hours, mostly overnight) on the weekend. I'll report back any issues I encounter.

esc commented:
> @jamesobutler thank you for the updates, I have submitted this to the Anaconda Internal Build Farm to smoketest (i.e. a short test, only a few hours) it. Since this is such a sweeping change I'd like to give it a full run (longer test, runs 12 hours, mostly overnight) on the weekend. I'll report back any issues I encounter.

All the existing 3.7 based builds fail as expected. And #8423 was merged.

esc commented:
@jamesobutler seems like this is now conflicting with `main` again... Would you be so kind and give this another rebase. I am hoping that, since all pre-cursors have been taken care of, that this will be the last rebase 🤞 -- and thank you again for your patience here 👍

jamesobutler commented:
I have removed my commit `Remove 32bit testing in CI` originally issued at https://github.com/numba/numba/pull/8319#discussion_r943924791 because later PR #8423 with the same type changes was integrated prior to this PR. This branch is now rebased with latest `main`.

Please advise about the following test that is only run for Python 3.7 which appears that it should be removed as well. This would be removal of the class `TestStencilBase` which is also inherited by `TestStencil` class.  This check to only run the test for Python 3.7 was added in https://github.com/numba/numba/commit/3c17d0706473a4a2fece2c3c30467df5e264ba30 when it was explicitly excluded for Python 3.9, but would've also begun to exclude Python 3.8 at that time as well.
https://github.com/numba/numba/blob/061f21a73a574ea1b3451910a36db4f9b3ae788b/numba/tests/test_stencils.py#L87-L88

stuartarchibald commented:
> Many thanks for raising the issue with the `@stencil` tests @jamesobutler. IIRC this was raised/discussed a while back when considering the support schedules for Python versions. I think this needs resolving ahead of dropping 3.7 support else part of the public API would effectively be untested. I'll raise this again at next weeks' maintainer's triage meeting and if not resolved also at the public Numba meeting. Thanks again!

The issue noted with `@stencil` tests is unblocked by #8518.

jamesobutler commented:
@stuartarchibald I've rebased this branch now that the stencil changes in #8518 have been integrated. I will keep an eye on the CI results here.

jamesobutler commented:
CI is green 🎉

esc commented:
@jamesobutler thank you, we'll commence reviewing this shortly.

esc commented on CHANGE_LOG:
```diff
@@ -1,3 +1,12 @@
+Version 0.57.0
```
We don't normally add items to the change log as part of the PR. But this doesn't do any harm so I don't see a reason to necessarily purge it. @stuartarchibald @sklam what is your take?

stuartarchibald replied:
Some thoughts (unordered):
1. I think that capturing notes on important changes in the change log as PRs are merged is a good idea. 
2. There's no policy on incrementally adding things to the change log and I think it would probably make sense to develop one ahead of doing something ad-hoc which might set a undesired precedence.
3. The general approach of writing to the top of the change log would potentially lead to more merge conflicts and this adds a burden on to contributors/maintainers to resolve them.
4. At present the release notes/change log are written based on the PRs list scraped from the git log/what's in the milestone (should be the same!). It would be unlikely an alteration such as minimum version change be missed (especially as the version support matrix would need to reflect this and validating that is part of the release process).
5. In future, I think having a system that assimilates important release notes from specially named/formatted individual files (or similar) may be good approach and something worth discussion, but it is unlikely such a system would be ready for this PR even if accepted as an idea.

Given "automatic production of release notes" has also been informally discussed via other channels I'll raise this at the Numba maintainers' meeting next week. In the interests of getting this PR merged and making the change set minimal, I think simply removing it would be best for now.

jamesobutler replied:
I added this to the changelog because this was something that you did @stuartarchibald in your commit that moved the minimum supported python version to 3.7 in https://github.com/numba/numba/commit/7bdd80023cc67a5896386ad8aa613ee26180b20f. I was trying to follow the best 😄 

I was also assuming it was because of 1) from above as it is a major change that people would be interested when developing numba along the `main` branch.

stuartarchibald replied:
> I added this to the changelog because this was something that you did @stuartarchibald in your commit that moved the minimum supported python version to 3.7 in [7bdd800](https://github.com/numba/numba/commit/7bdd80023cc67a5896386ad8aa613ee26180b20f). I was trying to follow the best smile. Seems like @sklam made a comment in agreement in [#6964 (comment)](https://github.com/numba/numba/pull/6964#discussion_r619423564).

Thanks for making the effort to form this patch based on an update of a prior similar change so as to make reviewing it easier, it's much appreciated! There indeed was an occurrence of an incremental write to the change log before, but I don't think it's happened since, therefore I would consider not doing it as having precedence. As noted above, this is of interest in relation to "automatic production of release notes" and will be discussed next week. If this something important to you then Numba also has weekly public meetings (details are pinned on Numba's discourse instance) where the schedule is open/community driven, you are welcome to raise it as a topic.

jamesobutler replied:
I've pushed a commit that removes the changelog diff

esc commented on azure-pipelines.yml:
```diff
@@ -12,8 +12,8 @@ jobs:
     matrix:
-      py37_np118:
-        PYTHON: '3.7'
+      py38_np118:
```
I see that all the 3.7 tests were simply ported to 3.8. Is this OK, or do we want to take some of these tests to 3.9 or 3.10 or are these potentially slots for 3.11? Maybe it is also OK to merge this as is and "re-balance" the distribution of Python versions across tests in a later PR. @sklam @stuartarchibald what do you think?

stuartarchibald replied:
Running the function `numba.tests.support.print_azure_matrix()` will print the Azure CI matrix as a table to `stdout`. Suggest looking at the difference between `main` and this PR first?

jamesobutler replied:
I manually counted up in this file where `main`
```
NumPy | Python | Count
-----------------------
 1.18 |  3.7   |   2
 1.18 |  3.8   |   1
 1.19.2=*_0 |  3.8   |   3
 1.19.2=*_0 |  3.9   |   1
 1.20 |  3.8   |   3
 1.20 |  3.9   |   1
 1.21 |  3.9   |   1
 1.21 |  3.10  |   1
 1.22 |  3.10  |   0
 1.22 |  3.8   |   1
 1.23 |  3.9   |   1
 1.23 |  3.10  |   2
```
and for this PR (all the 1.18/3.7 have been moved to 1.18/3.8):
```
NumPy | Python | Count
-----------------------
 1.18 |  3.8   |   3
 1.19.2=*_0 |  3.8   |   3
 1.19.2=*_0 |  3.9   |   1
 1.20 |  3.8   |   3
 1.20 |  3.9   |   1
 1.21 |  3.9   |   1
 1.21 |  3.10  |   1
 1.22 |  3.10  |   0
 1.22 |  3.8   |   1
 1.23 |  3.9   |   1
 1.23 |  3.10  |   2
```

stuartarchibald replied:
I think these changes are fine, thanks for checking.

esc commented on docs/environment.yml:
```diff
@@ -5,7 +5,7 @@ name: rtd
   - numba/label/dev
 dependencies:
-  - python=3.7
+  - python=3.8
```
Since this is for the docs build, we can take this straight to 3.10? Means we don't need to bump it every year.
```suggestion
  - python=3.10
```

stuartarchibald replied:
Is there any chance of the docs build somehow ending up with "newer" python features in the source which would stop docs building on the oldest supported Python?

jamesobutler replied:
@esc Would you like this bumped to 3.10? Or just keep it set to the minimum supported version as done in this PR currently?

esc replied:
@jamesobutler I don't know yet, there doesn't seem to be any consensus on this yet. Both options appear to have legitimate pros and cons.

jamesobutler replied:
@esc Since this PR continues the current idea of using the minimum supported version of python for building documentation, can we proceed with this PR as is? Switching to a newer python version 3.10 or 3.11 can happen in a separate PR.

esc replied:
@jamesobutler that sounds like a reasonable resolution, thank you.

esc commented on numba/_version.py:
```diff
@@ -46,8 +46,7 @@ def run_command(commands, args, cwd=None, verbose=False, hide_stderr=False):
     stdout = p.communicate()[0].strip()
-    if sys.version_info[0] >= 3:
-        stdout = stdout.decode()
+    stdout = stdout.decode()
```
The `.decode()` can be merged with the line above.

esc commented on numba/cloudpickle/cloudpickle.py:
```diff
@@ -65,21 +65,8 @@
 except ImportError:
     _typing_extensions = Literal = Final = None
 
-if sys.version_info >= (3, 5, 3):
```
This is a vendored cloudpickle, right? Maybe we should just retain their stuff verbatim rather than "forking" it? This will surely also make future cloudpickle updates easier? Maybe there is a version of cloudpickle already that has dropped 3.7 support?

jamesobutler replied:
Yes indeed starting with cloudpickle v2.2.0 they changed this line for python 3.8+ see https://github.com/cloudpipe/cloudpickle/blob/f31859b1dd83fa691f4f7f797166b262c9acb8e7/cloudpickle/cloudpickle.py#L68

Looks like numba’s cloudpickle has had a bug fix and typo fixed since the censored version was last updated to v1.6.0 see https://github.com/numba/numba/commits/main/numba/cloudpickle/cloudpickle.py

esc replied:
@jamesobutler I am not sure yet, I think @sklam is most experienced in the cloudpickle integration, so I would like to wait for their commentary before attempting to reach consensus.

sklam replied:
i prefer to handle cloudpickle separately.

esc replied:
Yes, please revert the changes to cloudpickle and submit a separate PR. Thank you for your effort.

jamesobutler replied:
Cloudpickle changes have now been removed from this PR.

esc reviewed (changes requested):
I took a look and most of it seems straightforward and almost mechanical. I've left some suggestions to be considered. Also, I am running this on the Anaconda internal build farm as: `numba_smoketest_cpu_yaml_142`.

stuartarchibald commented:
I'm surprised no C code needed changing. Perhaps grep for `PY_MINOR_VERSION`?

esc commented:
@jamesobutler thank you for your continued patience on this one. I am planning to begin reviewing this PR again next week. Would you be so kind as to fix the conflict and rebase this onto the current `main`.

jamesobutler commented:
@esc I have rebased this branch against current `main` and resolved the merge conflict.

esc commented:
`Build numba_smoketest_cpu_yaml_150 has started`

esc commented on numba/_dynfunc.c:
```diff
@@ -462,15 +448,11 @@ static PyTypeObject GeneratorType = {
-#if PY_MINOR_VERSION > 7
     0,                         /* tp_vectorcall */
-#if PY_MINOR_VERSION == 8
+/* Python 3.8 has two slots, 3.9 has one. */
```
Should this comment also indicate that everything beyond 3.9 also only has a single slot?

jamesobutler replied:
I’m not sure how much value the comment adds here beyond what is described by the actual code. If the comment is maintained it could also be shortened to just “Python 3.8 has two slots” and make no mention of other specific python versions.

esc replied:
@jamesobutler yes, shortening would also be an option. I am OK with that.

jamesobutler replied:
I've pushed an additional commit to shorten the comment. I've rebased to let automated testing run against latest `main` changes.

esc reviewed (changes requested):
I have given this another pass today and found one comment that could be improved.

jamesobutler commented:
Hi @esc, is there anything left to do here?

esc commented:
> Ok thanks for the update @esc. Are there any remaining changes from your end that are still being requested? If not, could you remove your "Changes requested" status for this PR?

Sure, done.

stuartarchibald commented on numba/cpython/hashing.py:
```diff
@@ -393,15 +370,9 @@ def impl(cgctx, builder, signature, args):
-        return impl
-    else:
-        def impl(val):
-            hashed = _Py_hash_t(_tuple_hash_resolve(val))
```
Is `_tuple_hash_resolve` now unused? I would guess that it is and so it could be removed entirely?

stuartarchibald commented on numba/cpython/randomimpl.py:
```diff
@@ -823,15 +820,7 @@ def _impl(alpha, beta):
 
-            if POST_PY38:
-                # Adjust due to cpython
-                # commit 63d152232e1742660f481c04a811f824b91f6790
```
Suggest keeping this comment.

stuartarchibald commented on numba/tests/test_analysis.py:
```diff
@@ -663,8 +663,7 @@ def _literal_const_sample_generator(self, pyfunc, consts):
         # create new code parts
         co_args = [pyfunc_code.co_argcount]
 
-        if utils.PYVERSION >= (3, 8):
-            co_args.append(pyfunc_code.co_posonlyargcount)
+        co_args.append(pyfunc_code.co_posonlyargcount)
         co_args.append(pyfunc_code.co_kwonlyargcount)
         co_args.extend([pyfunc_code.co_nlocals,
```
Now there's no need for a branch this can just be a direct declaration of the entire list?

stuartarchibald commented on numba/tests/test_parfors.py:
```diff
@@ -3363,8 +3363,7 @@ def generate_prange_func(self, pyfunc, patch_instance):
         # create new code parts
         co_args = [pyfunc_code.co_argcount]
 
-        if utils.PYVERSION >= (3, 8):
-            co_args.append(pyfunc_code.co_posonlyargcount)
+        co_args.append(pyfunc_code.co_posonlyargcount)
         co_args.append(pyfunc_code.co_kwonlyargcount)
         co_args.extend([pyfunc_code.co_nlocals,
                         pyfunc_code.co_stacksize,
```
This can just be a direct declaration of the list now there is no branch?

stuartarchibald commented on numba/tests/test_withlifting.py:
```diff
@@ -284,7 +284,6 @@ def test_liftcall4(self):
         msg = ("compiler re-entrant to the same function signature")
         self.assertIn(msg, str(raises.exception))
 
-    # 3.7 fails to interpret the bytecode for this example
```
Remove the:
```
 @unittest.skipIf(PYVERSION <= (3, 8),
                     "unsupported on py3.8 and before")
```
on the line below?

jamesobutler replied:
Introduced in https://github.com/numba/numba/commit/d5eb4ad50c4fb6ef504cd732c9a7960ab03cc22a, it says `test_liftcall5` should be skipped for Python 3.8 and earlier. Since we are still supporting Python 3.8, it would seem to make sense to keep this exclusion.  Are you suggesting that it actually shouldn't have been excluded when using Python 3.8?

stuartarchibald replied:
Thanks for pointing this out, will need to go and look at the bytecode, the comment and `@skipIf` suggest different things.

jamesobutler replied:
Based on this [force push](https://github.com/numba/numba/compare/2ea4fca71348b579a691bc4087de4d77fba0d582..d5eb4ad50c4fb6ef504cd732c9a7960ab03cc22a) it appears that it was originally thought to be 3.7 and earlier issue, but was actually a 3.8 and earlier issue. Would you like me to update the comment to say `# 3.8 fails to interpret the bytecode for this example`?

stuartarchibald replied:
I've manually checked this against Python 3.8. In Python 3.8 it fails in bytecode tracing/analysis as the same "state" is encountered twice. Think updating the comment would be good. Thanks.

jamesobutler replied:
I've added the comment back with the updated mention of 3.8 instead of 3.7.

stuartarchibald reviewed (commented):
@jamesobutler thanks for your efforts in refactoring/code removal to move the Python baseline version to 3.8. I've given this a review and there's a few things to address inline, a couple of comments might need input from a maintainer as the fix could be involved. Further, there are a few things that are need addressing which I'll leave in subsequent comments. Thanks again!

esc commented:
@jamesobutler thank you again for your hard work and sustained patience on this issue! I just realized that llvmlite will need this change. If you do have some spare capacity, I have opened the ticket at: https://github.com/numba/llvmlite/issues/891 :: that would be terrific!! If not, that is fine too, no hard feelings.

jamesobutler commented:
> - https://github.com/numba/numba/blob/036d9561dc4188222febadb3fcb621519e241990/docs/source/user/troubleshoot.rst?plain=1#L1044-L1078
> the above code-block needs to refer to swap `python3.7` for `python3.10` or similar, and then e.g. a `cpython-310` DSO name used in
> https://github.com/numba/numba/blob/036d9561dc4188222febadb3fcb621519e241990/docs/source/user/troubleshoot.rst?plain=1#L1078

@stuartarchibald Do you mind providing the updated output example for this? I can either commit it after you post a comment with it or feel free to push a commit directly to my branch here with the change. I don't work on Linux platforms and would like to get the full updated output rather than trying to fake it by updating parts of the outputs with `3.7`->`3.10` changes.

jamesobutler commented:
> - https://github.com/numba/numba/blob/036d9561dc4188222febadb3fcb621519e241990/numba/core/controlflow.py#L794-L796
> can be removed, as can any bytecode handling involving SETUP_EXCEPT and SETUP_FINALLY (suggest greping for those strings) and tests involving these op codes.

I can understand removing SETUP_EXCEPT handling because it was removed starting in Python 3.8 (https://docs.python.org/3/whatsnew/3.8.html#cpython-bytecode-changes), but SETUP_FINALLY appears to still be present and only removed starting in Python 3.11 (https://docs.python.org/3/whatsnew/3.11.html#changed-removed-opcodes). Shouldn't testing for SETUP_FINALLY be maintained?

jamesobutler commented on numba/tests/test_ir_inlining.py:
```diff
@@ -438,42 +438,6 @@ def bar(x):
 
         self.check(impl, inline_expect={'foo': True}, block_count=1)
 
-    @skip_py38_or_later
```
This test was originally skipped (https://github.com/numba/numba/commit/05cce6a6110876efbbd807f57270e16c9f41e9ed) starting when Python 3.8 support was added.

From #4755, it is unclear if this looplifting issue was ever resolved in a later PR
> Remaining problems:
> 
> - [ ] tuple hash algo changed. https://bugs.python.org/issue34751
> - [ ] looplifting is broken due to CFG assumption changed.
> _Edit_ The fixes for the above unresolved py3.8 problems will be splitted to separate PRs. This PR now contains the necessary changes to make Py3.8 mostly working.

jamesobutler replied:
I have brought in your commit from that PR into the PR here. It is listed as the first commit.

stuartarchibald commented:
> > * https://github.com/numba/numba/blob/036d9561dc4188222febadb3fcb621519e241990/docs/source/user/troubleshoot.rst?plain=1#L1044-L1078
> >   
> >   the above code-block needs to refer to swap `python3.7` for `python3.10` or similar, and then e.g. a `cpython-310` DSO name used in
> >   https://github.com/numba/numba/blob/036d9561dc4188222febadb3fcb621519e241990/docs/source/user/troubleshoot.rst?plain=1#L1078
> 
> @stuartarchibald Do you mind providing the updated output example for this? I can either commit it after you post a comment with it or feel free to push a commit directly to my branch here with the change. I don't work on Linux platforms and would like to get the full updated output rather than trying to fake it by updating parts of the outputs with `3.7`->`3.10` changes.

@jamesobutler I've taken a look at this and what's generated. The example is quite verbose now due to various other changes in the way Numba represents things in DWARF, so I think maybe just leave it for now and it can be addressed later. It's not invalid as it is, it's just the output contents has changed, but the reality is that most of the changes to output come from other updates to Numba and not just 3.7->3.10.

stuartarchibald commented on docs/source/user/cli.rst:
```diff
@@ -82,61 +82,192 @@ Example output::
+    psutil                    5.9.0           py310hca72f7f_0    defaults
+    ptyprocess                0.7.0              pyhd3eb1b0_2    defaults
+    pure_eval                 0.2.2              pyhd3eb1b0_0    defaults
+    pycparser                 2.21               pyhd3eb1b0_0    defaults
+    pygments                  2.11.2             pyhd3eb1b0_0    defaults
+    pyparsing                 3.0.9           py310hecd8cb5_0    defaults
+    python                    3.10.8               h218abb5_1    defaults
+    python-dateutil           2.8.2              pyhd3eb1b0_0    defaults
+    pyyaml                    6.0                      pypi_0    pypi
+    pyzmq                     23.2.0          py310he9d5cce_0    defaults
+    readline                  8.2                  hca72f7f_0    defaults
+    scipy                     1.9.3           py310h09290a1_0    defaults
+    setuptools                65.5.0          py310hecd8cb5_0    defaults
+    six                       1.16.0             pyhd3eb1b0_1    defaults
+    smmap                     4.0.0              pyhd3eb1b0_0    defaults
+    sqlite                    3.40.0               h880c91c_0    defaults
+    stack_data                0.2.0              pyhd3eb1b0_0    defaults
+    tapi                      1000.10.8            ha1b3eb9_0    defaults
+    tbb                       2021.5.0             haf03e11_0    defaults
+    tbb-devel                 2021.5.0             haf03e11_0    defaults
+    tk                        8.6.12               h5d9f67b_0    defaults
+    tornado                   6.2             py310hca72f7f_0    defaults
+    traitlets                 5.1.1              pyhd3eb1b0_0    defaults
+    typing-extensions         4.3.0           py310hecd8cb5_0    defaults
+    typing_extensions         4.3.0           py310hecd8cb5_0    defaults
+    tzdata                    2022f                h04d1e81_0    defaults
+    wcwidth                   0.2.5              pyhd3eb1b0_0    defaults
+    wheel                     0.37.1             pyhd3eb1b0_0    defaults
+    xz                        5.2.6                hca72f7f_0    defaults
+    yaml                      0.2.5                haf1e3a3_0    defaults
+    zeromq                    4.3.4                h23ab428_0    defaults
+    zipp                      3.8.0           py310hecd8cb5_0    defaults
+    zlib                      1.2.13               h4dc903c_0    defaults
+
+    No errors reported.
+
+
+    __Warning log__
+    Warning (cuda): CUDA driver library cannot be found or no CUDA enabled devices are present.
+    Exception class: <class 'numba.cuda.cudadrv.error.CudaSupportError'>
```
```suggestion
    (output truncated due to length)
```
Suggest putting this in similarly to how it was originally, not sure that adding all the packages is useful for a sample output.

stuartarchibald commented on numba/mviewbuf.c:
```diff
@@ -294,11 +294,7 @@ static PyTypeObject MemAllocType = {
     0,                                          /* tp_getattr */
     0,                                          /* tp_setattr */
-#if PY_MAJOR_VERSION >= 3
     0,                                          /* tp_reserved */
```
I think the second part of this comment still needs addressing as of 8871142?

jamesobutler replied:
>  and in Python 3.10+ a tp_vectorcall entry exists following the tp_finalized. (applicable to all cases of PyTypeObject type declarations).

Are you suggesting there is something missing/incorrect regarding this code here for Python 3.10 support and numba already supports Python 3.10? Shouldn't that be a separate PR as this is just addressing dropping Python 3.7 support?

stuartarchibald replied:
Agree, will fix later.

stuartarchibald replied:
xref #8666

stuartarchibald commented on numba/core/utils.py:
```diff
@@ -375,68 +375,7 @@ def __setitem__(self, key, value):
 
-
-if PYVERSION > (3, 7):
     from functools import cached_property
```
This still needs addressing as of ed973a2?

jamesobutler replied:
This has been addressed by @esc

stuartarchibald commented:
> * https://github.com/numba/numba/blob/036d9561dc4188222febadb3fcb621519e241990/numba/core/serialize.py#L171-L172 suggests it can be removed.

I think ^ may still need addressing as of ed973a2? Many thanks.

jamesobutler commented:
> Note to self: have reviewed up to https://github.com/numba/numba/commit/ed973a2577449aff52aa06a80925a0847e06ddae

[force push](https://github.com/numba/numba/compare/ed973a2577449aff52aa06a80925a0847e06ddae..7abbb973a47dd093e5e8ade55849aee8f095c81d) has squashed down previous fix me up commits. There is no diff change.

[force push](https://github.com/numba/numba/compare/7abbb973a47dd093e5e8ade55849aee8f095c81d..79f02648bcb0f75316c5da04579beb8227acb097) rebases against latest `main` to pull in all recent changes for automated test cases to run with.

[force push](https://github.com/numba/numba/compare/79f02648bcb0f75316c5da04579beb8227acb097..6e4cebbc8f7b05dce64db7e4758927d334b9f939) brings in @stuartarchibald's commit from #8651 as suggested in https://github.com/numba/numba/pull/8319#discussion_r1042123851. An additional commit has been pushed which drops the python 3.7 support from the changes in @stuartarchibald's commit from #8651.

jamesobutler commented:
Hi @esc I was aware of the force push concerns that is why I clearly described what I had done in https://github.com/numba/numba/pull/8319#issuecomment-1341538850 with all the links to see the diffs of each force push. This allowed me to clearly see that tests would continue to pass when this high priority PR is ultimately merged. I’m trying to be very diligent about the overall health of this branch and how it will be when merged into `main`.

From my GitHub profile you can see that I’m located in the EST time zone and I typically work on this later in the day. What time zones are you all located in? Then I will be able to know when your cut off is when you mention “by the end of this week”. Also as maintainers I still have the PR option checked to allow you to push directly to this branch if you would like to make changes to then integrate this PR before I get to them.

esc commented:
> Hi @esc I was aware of the force push concerns that is why I clearly described what I had done in [#8319 (comment)](https://github.com/numba/numba/pull/8319#issuecomment-1341538850) with all the links to see the diffs of each force push. This allowed me to clearly see that tests would continue to pass when this high priority PR is ultimately merged. I’m trying to be very diligent about the overall health of this branch and how it will be when merged into `main`.

OK. I understand. Just bear in mind that what makes it easier for you may make it harder for us.

> 
> From my GitHub profile you can see that I’m located in the EST time zone and I typically work on this later in the day. What time zones are you all located in? Then I will be able to know when your cut off is when you mention “by the end of this week”. Also as maintainers I still have the PR option checked to allow you to push directly to this branch if you would like to make changes to then integrate this PR before I get to them.

I can't speak for the others but I think we are spread across the globe. Ideally it should be merged during the merge-party on Friday, which is usually in the US-evening. So ideally we want to get the last bits reviewed and ready for merge during EU daytime tomorrow. If you can't make it today, that is fine too and thank you for the explicit permission to push to your branch, we may very well make use of this.

stuartarchibald commented:
PR #8660 (abandoned) captures the history up to ed973a2577449aff52aa06a80925a0847e06ddae (as referenced above). Hopefully this will help with re-reviewing across the force pushes.

stuartarchibald commented:
@esc Locally, I've merged 0441bb1 (HEAD of `main` on 6th Dec 2022) into #8660 and compared it to 017a64e, the diff is essentially the patch #8651 with a minor modification to drop the Python 3.7 specific branch.

esc commented:
In order to check that the squashed commit, which is now 6e4cebbc8f7b05dce64db7e4758927d334b9f939 matches the previous history at #8660, I did the following

First, generate the diff that should be created by the squash:

`git diff fb952f8f5959ce31583da612a91b204bca58da81..wip/preserve_pr_8319 > PRE-REBASE`

Then, generate the diff for the squashed commit 

`git diff "6e4cebbc8f7b05dce64db7e4758927d334b9f939^..6e4cebbc8f7b05dce64db7e4758927d334b9f939" > POST-REBASE`

The `diff PRE-REBASE POST-REBASE`,

and I get:

```
 💣 zsh» diff PRE-REBASE POST-REBASE                                                                                                                                                                                                            :(
1235c1235
< index 6816cc4e18..6ccb86d1ae 100644
---
> index 2bc70e2765..472fa01c8b 100644
1344c1344
< index 423c4ec5e4..faf8a20ea4 100644
---
> index 0202d8b8cd..38d6fe3c36 100644
1356c1356
< @@ -3188,7 +3188,7 @@ def make_nditer_cls(nditerty):
---
> @@ -3198,7 +3198,7 @@ def make_nditer_cls(nditerty):
1365c1365
< @@ -3280,7 +3280,7 @@ def make_nditer_cls(nditerty):
---
> @@ -3290,7 +3290,7 @@ def make_nditer_cls(nditerty):
1565,1620d1564
< diff --git a/numba/tests/test_ir_inlining.py b/numba/tests/test_ir_inlining.py
< index ce5909a732..229e13339c 100644
< --- a/numba/tests/test_ir_inlining.py
< +++ b/numba/tests/test_ir_inlining.py
< @@ -28,7 +28,7 @@ from numba.core.cpu import InlineOptions
<  from numba.core.compiler import DefaultPassBuilder, CompilerBase
<  from numba.core.typed_passes import InlineOverloads
<  from numba.core.typing import signature
< -from numba.tests.support import (TestCase, unittest, skip_py38_or_later,
< +from numba.tests.support import (TestCase, unittest,
<                                   MemoryLeakMixin, IRPreservingTestPipeline,
<                                   skip_parfors_unsupported,
<                                   ignore_internal_warnings)
< @@ -438,42 +438,6 @@ class TestFunctionInlining(MemoryLeakMixin, InliningBase):
<
<          self.check(impl, inline_expect={'foo': True}, block_count=1)
<
< -    @skip_py38_or_later
< -    def test_inline_involved(self):
< -
< -        fortran = njit(inline='always')(_gen_involved())
< -
< -        @njit(inline='always')
< -        def boz(j):
< -            acc = 0
< -
< -            def biz(t):
< -                return t + acc
< -            for x in range(j):
< -                acc += biz(8 + acc) + fortran(2., acc, 1, 12j, biz(acc))
< -            return acc
< -
< -        @njit(inline='always')
< -        def foo(a):
< -            acc = 0
< -            for p in range(12):
< -                tmp = fortran(1, 1, 1, 1, 1)
< -
< -                def baz(x):
< -                    return 12 + a + x + tmp
< -                acc += baz(p) + 8 + boz(p) + tmp
< -            return acc + baz(2)
< -
< -        def impl():
< -            z = 9
< -
< -            def bar(x):
< -                return foo(z) + 7 + x
< -            return bar(z + 2)
< -
< -        self.check(impl, inline_expect={'foo': True, 'boz': True,
< -                                        'fortran': True}, block_count=37)
< -
<      def test_inline_renaming_scheme(self):
<          # See #7380, this checks that inlined variables have a name derived from
<          # the function they were defined in.
1622c1566
< index f9dc74dacb..95e2e2315e 100644
---
> index a321c5b4c4..f1f9367534 100644
1633c1577
< @@ -1454,32 +1453,12 @@ class TestJitClassOverloads(MemoryLeakMixin, TestCase):
---
> @@ -1466,32 +1465,12 @@ class TestJitClassOverloads(MemoryLeakMixin, TestCase):
```

So I guess that is mostly fine. Except `for the test_ir_inling` which exists only in PRE-REBASE?

esc commented:
> So I guess that is mostly fine. Except `test_ir_inling` which exists only in PRE-REBASE?

I see, this was originally deleted, but in the new variant the test is changed instead by both 017a64e684a6da16cc5653c551bc5735c5e54d53 and 5da4bb854cd4309cdc9724f3dd21ebc9569b6ff6

jamesobutler commented:
Yes you all are describing what I mentioned in https://github.com/numba/numba/pull/8319#issuecomment-1341538850. Before I did those changes there was not @stuartarchibald’s `test_ir_inlining` changes from #8651. Though more details are in that comment, again I’ll reiterate I squashed down what I previously had as captured in #8660, rebased against main to pull in latest commits for testing, and rebased to include @stuartarchibald’s commit from #8651. That commit became the first commit in the PR because it is Python 3.7+ compatible and resolved merge conflicts of what I had previously simply removed entirely. Then I pushed an additional commit here as seen as the 3rd commit which removes the 3.7 support from @stuartarchibald’s commit as it after the main commit in the git history that drops python 3.8 support.

jamesobutler commented:
It appears that with recent pushes to `main` this branch needs to be rebased again to resolve current merge conflicts that are now present. @esc I have no problem handling that, but I will wait to proceed on your orders.

esc commented:
I have pushed 24c72a2c3c7244a400b38de6e611cdd214e859aa which fixed the stuff in `serialize.py` as requested by: https://github.com/numba/numba/pull/8319#issuecomment-1340804271

esc commented:
I have pushed 0afb22056e5315c429da6969cb508e357526167d -- this fixes the last item on the list:  https://github.com/numba/numba/pull/8319#discussion_r1042061871 -- I had to remove the `PYVERSION` import in that one, since it was no longer being used in the module.

esc commented:
> I have pushed [0afb220](https://github.com/numba/numba/commit/0afb22056e5315c429da6969cb508e357526167d) -- this fixes the last item on the list: [#8319 (comment)](https://github.com/numba/numba/pull/8319#discussion_r1042061871) -- I had to remove the `PYVERSION` import in that one, since it was no longer being used in the module.

I had to partially revert 0afb220 , the import was actually a re-export.

esc commented:
I pushed [3cc6322](https://github.com/numba/numba/pull/8319/commits/3cc63220e2752ba3cacd580bfa5c6cf0bf1ab210)  to fix the conflicts with `main` -- this seems to have worked as Azure has been triggered!

esc commented:
> I pushed [3cc6322](https://github.com/numba/numba/pull/8319/commits/3cc63220e2752ba3cacd580bfa5c6cf0bf1ab210) to fix the conflicts with `main` -- this seems to have worked as Azure has been triggered!

This was not sufficient, it resolved the conflict but the code broke. With 45d4e452aed00c008c1009d0d510c7daedd85d0a I have merged in `main` to resolve this.

stuartarchibald commented:
Have checked 24c72a2, 0afb220 and f4bc00a, these cover the outstanding items on the review.
Have checked 3cc6322 and 45d4e45 as resolving conflict(s) against `main`.

Thanks for the patches @esc .

esc commented:
I have scheduled this to run on the Anaconda Internal buildfarm as `numba_smoketest_cpu_yaml_156`.

stuartarchibald reviewed (approved):
Many thanks for all your work/efforts on this @jamesobutler and @esc. Am glad to have this maintenance task completed!

esc commented:
> I have scheduled this to run on the Anaconda Internal buildfarm as `numba_smoketest_cpu_yaml_156`.

This was green, as was Azure.
